Python/specialize.c

```diff
@@ -637,3 +640,103 @@ _Py_Specialize_LoadGlobal(
+    PyObject **stack_pointer, uint8_t original_oparg,
+    _Py_CODEUNIT *instr, SpecializedCacheEntry *cache)
+{
+    PyObject *callable = stack_pointer[-(original_oparg + 1)];
+    _PyAdaptiveEntry *cache0 = &cache->adaptive;
+    PyTypeObject *type = Py_TYPE(callable);
+    /* Specialize C functions */
+    if (PyCFunction_CheckExact(callable)) {
+        PyCFunctionObject *meth = (PyCFunctionObject *)callable;
+        if (PyCFunction_GET_FUNCTION(callable) == NULL) {
+            goto fail;
+        }
+        /* Currently not optimizing anything that is FASTCALL, has keywords, has
+           varargs, or has no args. Microbenchmarks show they don't benefit much to be
+           worth a specialized instruction.
+        */
+        switch (PyCFunction_GET_FLAGS(meth) & (METH_VARARGS | METH_FASTCALL |
+            METH_NOARGS | METH_O | METH_KEYWORDS | METH_METHOD)) {
+            case METH_FASTCALL:
+                SPECIALIZATION_FAIL(CALL_FUNCTION, type, callable,
+                    "_PYCFUNCTION_FAST");
+                goto fail;
+            case METH_O:
+                // PYCFUNCTION_O;
+                *instr = _Py_MAKECODEUNIT(CALL_CFUNCTION_O, _Py_OPARG(*instr));
+                goto success;
+            case METH_VARARGS:
+                SPECIALIZATION_FAIL(CALL_FUNCTION, type, callable, "PYCFUNCTION");
+                goto fail;
+            case METH_VARARGS | METH_KEYWORDS:
+                SPECIALIZATION_FAIL(CALL_FUNCTION, type, callable, 
+                    "PYCFUNCTION_WITH_KEYWORDS");
+                goto fail;
+            case METH_FASTCALL | METH_KEYWORDS:
+                SPECIALIZATION_FAIL(CALL_FUNCTION, type, callable, 
+                    "_PYCFUNCTION_FAST_WITH_KEYWORDS");
+                goto fail;
+            case METH_NOARGS:
+                SPECIALIZATION_FAIL(CALL_FUNCTION, type, callable, "PYCFUNCTION_NOARGS");
+                goto fail;
+            /* This case should never happen with PyCFunctionObject -- only
+               PyMethodObject. See zlib.compressobj()'s methods for an example.
+            */
+            case METH_METHOD | METH_FASTCALL | METH_KEYWORDS:
+                // PYCMETHOD
+            default:
+                SPECIALIZATION_FAIL(CALL_FUNCTION, type, callable, "bad call flags");
+                goto fail;
+        }
+    }
+    /* These will be implemented in the future. Collecting stats for now. */
+#if SPECIALIZATION_STATS
+    if (PyFunction_Check(callable)) {
+        SPECIALIZATION_FAIL(CALL_FUNCTION, type, callable, "python function");
+        goto fail;
+    }
+    if (PyInstanceMethod_Check(callable)) {
+        SPECIALIZATION_FAIL(CALL_FUNCTION, type, callable, "new style bound method");
+        goto fail;
+    }
+    if (PyMethod_Check(callable)) {
+        SPECIALIZATION_FAIL(CALL_FUNCTION, type, callable, "bound method");
+        goto fail;
+    }
+    if (PyType_Check(callable)) {
+        if ((PyObject_HasAttrString(callable, "__dict__") ||
```

Comment:
Just split this into two categories for now. Immutable and mutable classes.
 `PyType_HasFeature(type, Py_TPFLAGS_IMMUTABLETYPE)`
We can refine it later if useful.

Reply:
Honestly I don't remember when ``Py_TPFLAGS_IMMUTABLETYPE`` is set for classes (I missed the entire immutable heap types mega-issue cleanup by Erlend on bpo), so I'll do more research on it in the future. So please tell me if something seems off. Thanks.

Reply:
Sorry, I put my comment on the wrong line. I was suggesting that you don't attempt to determine if a class is a Python class or not, just whether it is immutable. No need to look in `builtins`.
```C
if (PyType_Check(callable)) {            
    SPECIALIZATION_FAIL(
        CALL_FUNCTION, type, callable, 
        PyType_HasFeature(type, Py_TPFLAGS_IMMUTABLETYPE) ?
        "immutable class" :  "mutable class"
    );
}
```

Reply:
Ah alright. Thanks.